### PR TITLE
Fix RawStateDelta computation for overflowing big ints and floats

### PR DESCRIPTION
**Pull request by t0yv0**
The test suite in pulumi-gcp points out a corner case that was not handled previously and may lead to:

    error: recovered raw state does not byte-for-byte match the original raw state

This is the case of overflowing numbers (ints and floats) that cannot fit precisely in the float64 representation used by Pulumi. If the bridged provider maintainer overrides the bridge metadata to project these numbers as strings in the SDKs, everything works as expected. However there are practical cases where this maintainer action has not been taken yet; one reason to defer it is that changing the type from Number to String is a breaking change on the SDKs and may need to await a major version.

The change tightens equality to be stricter than equal-up-to-float64 precision, and if Pulumi representation does not equal the TF representation exactly, the provider will now backtrack and write the exact TF representation into the Raw State Delta, instructing the reader to replace the Pulumi representation with the TF representation.

See also: https://github.com/pulumi/pulumi-gcp/pull/3212

### CODE REVIEW

File: pkg/tfbridge/rawstate_test.go [hunk 1]
```diff
@@ -1250,3 +1279,71 @@ func Test_rawStateDelta_PropertyValue_serialization(t *testing.T) {
...
+			similar: true,
+		},
+		{
+			f64:    2495055709147741000,
+			bigNum: "2495055709147741188",
+			// Rounding occurred on something that could be an ID; must mark as similar: false.
+			similar: false,
+		},
+		{
+			f64:    2495055709147741188,
```
Comment (VenelinMartinov) [inline pkg/tfbridge/rawstate_test.go:1320]: I'm confused about this one - is the difference the types? Why are they not equal?

File: pkg/tfbridge/rawstate_test.go [hunk 2]
```diff
@@ -1250,3 +1279,71 @@ func Test_rawStateDelta_PropertyValue_serialization(t *testing.T) {
...
+			bigNum: "2495055709147741188",
+			// Rounding occurs here as well as the literal does not represent cleanly in float64.
+			similar: false,
+		},
+		{
+			f64:    2495055709147741000.32,
+			bigNum: "2495055709147741000.32",
+			// There is rounding here but the literal is not an integer and unlikely an ID; either value of
+			// similar: false or similar: true may be acceptable.
+			similar: false,
```
Comment (VenelinMartinov) [inline pkg/tfbridge/rawstate_test.go:1330]: Also this case - why are these not similar?

File: pkg/tfbridge/rawstate_test.go [hunk 3]
```diff
@@ -1250,3 +1279,71 @@ func Test_rawStateDelta_PropertyValue_serialization(t *testing.T) {
...
+		},
+		{
+			f64: 2495055709147741000.123456,
+			// This accidentally parses as an integer in big.Float;
```
Comment (VenelinMartinov) [inline pkg/tfbridge/rawstate_test.go:1334]: Can you please also explain this point? How can this parse as an integer?

**Review by VenelinMartinov — COMMENTED** [review]
Is it true that this change means that more cases considered "equal" in the raw state comparison will now be "not equal", so will now have a Replace Delta?
Are any cases affected in the inverse way? Is this strict:

>The change tightens equality to be stricter than equal-up-to-float64 precision 

Can you please explain some of the test cases, specifically the one which says that the float will be parsed as a int? Why is that fine?

File: pkg/tfbridge/rawstate_test.go (hunk 3 above)
Comment (t0yv0) [reply pkg/tfbridge/rawstate_test.go:1342]: I added the result of parsing to big.Float and stringifying as integer, hope it clarifies this behavior.

File: pkg/tfbridge/rawstate_test.go (hunk 1 above)
Comment (t0yv0) [reply pkg/tfbridge/rawstate_test.go:1326]: f64 cannot represent 2495055709147741188 so it truncates.

File: pkg/tfbridge/rawstate_test.go (hunk 2 above)
Comment (t0yv0) [reply pkg/tfbridge/rawstate_test.go]: Because `bigNumStr: autogold.Expect("2495055709147741000.2")`
